### PR TITLE
dialects: (x86) PR3 - Assembly emission

### PR DESCRIPTION
**Pull request by KGrykiel**
I've re-added the printing functionality with a filecheck to test it. I have temporarily brought back the get_register method to get the PR out, so please let me know what's the better way to do it. test.op no longer works when a particular register needs to be specified (it fails an assertion error during the printing). Riscv did it by starting off with operations that don't need register arguments, but I don't have those yet.

### CODE REVIEW

File: tests/filecheck/dialects/x86/x86_assembly_emission.mlir [hunk 1]
```diff
@@ -0,0 +1,7 @@
+// RUN: xdsl-opt -t x86-asm %s | filecheck %s
+
+%0 = x86.get_register : () -> !x86.reg<rax>
+%1 = x86.get_register : () -> !x86.reg<rdx>
```
Comment (superlopuh) [inline tests/filecheck/dialects/x86/x86_assembly_emission.mlir:4]: Please use "test.op" instead
```suggestion
%0, %1 = "test.op"() : () -> (!x86.reg<rax>, !x86.reg<rdx>)
```

Comment (superlopuh) [reply tests/filecheck/dialects/x86/x86_assembly_emission.mlir:4]: With this, the GetRegister op is not needed for this PR

File: xdsl/dialects/x86/ops.py [hunk 2]
```diff
@@ -167,3 +208,83 @@ class AddOp(RROperation[GeneralRegisterType, GeneralRegisterType]):
...
+    name: str,
+    arg_str: str,
+    comment: StringAttr | None = None,
+    is_indented: bool = True,
+) -> str:
+    code = "    " if is_indented else ""
+    code += name
+    if arg_str:
+        code += f" {arg_str}"
+    code = _append_comment(code, comment)
```
Comment (superlopuh) [inline xdsl/dialects/x86/ops.py:250]: I'd recommend moving the assembly printing helpers to `assembly.py`

**Comment by webmiche** [comment]
> test.op no longer works when a particular register needs to be specified (it fails an assertion error during the printing). 

I am very surprised by that. You "should" be able to generate SSAValues with any type through the test.op. What is your error?

**Comment by KGrykiel** [comment]
> > test.op no longer works when a particular register needs to be specified (it fails an assertion error during the printing).
> 
> I am very surprised by that. You "should" be able to generate SSAValues with any type through the test.op. What is your error?

I'm getting an error at the assertion at line 256:
`        assert isinstance(op, X86Op), f"{op}"`
it says that test_op is not a X86Op. Removing the asserion doesn't help either as the rest of the method stops working due to there not being a "assembly_line" method for a class of type "Operation"

**Comment by webmiche** [comment]
Ah, I see. So you essentially want to print these things using your own printing assembly infrastructure and you do not have the test.op in there.

Other classic suggestions are wrapping this in a block with blockarguments to get the SSAValues from there, I am not sure this will work in this case though.

**Review by superlopuh — APPROVED** [review]
My bad for not reading the PR description. I see the issue. I think in that case we might as well add it.

File: xdsl/dialects/x86/ops.py (hunk 2 above)
Comment (superlopuh) [reply xdsl/dialects/x86/ops.py:250]: no problem, I'll file an issue

**Comment by KGrykiel** [comment]
@webmiche @tarinduj is it fine if I merge and move on to the next PR?

File: xdsl/dialects/x86/ops.py (hunk 2 above)
Comment (superlopuh) [reply xdsl/dialects/x86/ops.py:250]: #2377

**Review by tarinduj — APPROVED** [review]

**Comment by webmiche** [comment]
> @webmiche @tarinduj is it fine if I merge and move on to the next PR?

Sure, I am fine for the moment. I think we should find a more principled way to model registers and global state instead fo spawning them from thin air, but that feels like a more general refactor than what this PR attempts. Feel free to merge!

**Pull request merged**